idarling/interface/painter.py

```diff
@@ -10,300 +10,148 @@
 import ida_kernwin
-import ida_nalt
-import ida_registry
-import idautils
+
+from PyQt5.QtCore import (  # noqa: I202
+    QAbstractItemModel,
+    QModelIndex,
+    QObject,
+    Qt,
+)
+from PyQt5.QtGui import QColor
+from PyQt5.QtWidgets import QStyledItemDelegate, QWidget
+import sip
+
+from .widget import StatusWidget
 
 if sys.version_info > (3,):
     long = int
 
 
-class Painter(object):
-    """The painter module is responsible for all the database painting."""
+class Painter(QObject):
+    class ProxyItemDelegate(QStyledItemDelegate):
+        def __init__(self, delegate, parent=None):
+            super(Painter.ProxyItemDelegate, self).__init__(parent)
+            self._delegate = delegate
 
-    DEFCOLOR = 0xffffffff
+        def paint(self, painter, option, index):
+            index = self.parent().model.index(index.row(), index.column())
+            self._delegate.paint(painter, option, index)
 
-    @staticmethod
-    def _get_ida_bg_color():
-        """
-        Return the background color from the disassembly view. It uses the
-        IDA registry to recover the color palettes configured by the user. Then
-        it loads the background color from the currently active palette.
-        """
-        palette = ida_registry.reg_read_binary("Palette")
-        if palette is None:
-            return 0xffffff
-        selected = struct.unpack("<I", palette[8:12])[0]
-        index = 176 + selected * 208
-        return struct.unpack("<I", palette[index : index + 4])[0]  # noqa: E203
+    class ProxyItemModel(QAbstractItemModel):
+        def __init__(self, model, parent=None):
+            super(Painter.ProxyItemModel, self).__init__(parent)
+            self._model = model
+            self._plugin = parent._plugin  # FIXME: Don't do this
 
-    @staticmethod
-    def _get_instruction_color(address):
-        """This is wrapper around get_item_color."""
-        return ida_nalt.get_item_color(address)
+        def index(self, row, column, parent=QModelIndex()):
+            return self.createIndex(row, column)
 
-    @staticmethod
-    def _set_instruction_color(address, color):
-        """This is a wrapper around set_item_color."""
-        ida_nalt.set_item_color(address, color)
+        def parent(self, index):
+            index = self._model.index(index.row(), index.column())
+            return self._model.parent(index)
 
-    @staticmethod
-    def _get_function_color(function):
-        return function.color
+        def rowCount(self):  # noqa: N802
+            return self._model.rowCount()
 
-    @staticmethod
-    def _set_function_color(function, color):
-        function.color = color
-        ida_funcs.update_func(function)
+        def columnCount(self):  # noqa: N802
+            return self._model.columnCount()
 
-    @staticmethod
-    def _refresh_navbar():
-        """Request a refresh of the navbar."""
-        ida_kernwin.refresh_navband(True)
+        def data(self, index, role=Qt.DisplayRole):
```

Comment:
Qt.DisplayRole seems to be not necessary
def data(self, index, role):

Reply:
It is. We're not calling `data()`, in which case it would have been. Here we're overriding the method, so we should keep the default value of arguments.